### DISCUSSION TITLE
Developing help

### DISCUSSION THREAD

**Discussion by Vitus-TL** [fedorovvl/tso_client#306] [Q&A]
Hi,

I'm quite new to AS3 scripting (but have some minor experience with embedded C/C++/Bash/Assemble). One of the issue's I encounter, is how to get a debug log. I tried the following code, but it's rejected:
//package LOG
//{
//      import { ExternalInterface.call } from 'flash.external.ExternalInterface';
//      ExternalInterface.call = require( 'flash.external.ExternalInterface');
//}
And although my current script checks out with: node --check, it has bad interaction on the workings of the TSO client.
Is it true the user scripts are run, before opening the menu? And what can I do, to limit bad effects to the client due to bad coding?
Also, I'm trying to use the function GetRemainingConstructionDuration() from GO/cBuilding.as but I can't figure out how to normalize the output in percentage:
       public function GetRemainingConstructionDuration():int
        {
            var _local_1:Number = 0;
            switch (this.mBuildingMode)
            {
                case cBuilding.BUILDING_MODE_SET_BUILDING_GROUND_PLACE:
                case cBuilding.BUILDING_MODE_SETTLER_WALKS_TO_BUILDING_GROUND_PLACE:
                    if (((!(this.GetResourceCreation() == null)) && (!(this.GetResourceCreation().GetPath() == null))))
                    {
                        _local_1 = (_local_1 + (((this.GetResourceCreation().GetPath().pathLenX10000 - this.GetResourceCreation().pathPos) / cComputeResourceCreation.SETTLER_WALK_SPEED_INT) / mGeneralInterface.mGlobalTimeScale));
                    };
                case cBuilding.BUILDING_MODE_CONSTRUCTION:
                    if (this.GetGOContainer() != null)
                    {
                        _local_1 = (_local_1 + ((this.GetGOContainer().mConstructionDuration * 1000) * (1 - (this.mBuildingProgress / (100 * defines.BUILDING_PROGRESS_SCALE_FACTOR)))));
                    };
                    break;
            };
            return (int(_local_1));
        }

Can you help me, with these question?

Cheers,
Emanuel (aka TL)

**Comment by fedorovvl** [comment]
hi.
js engine in adobe air supports only es5.. so u cant write scripts like this =) use pure js
there is debug function in client.. just write debug("something") and check introspector console
ab duration - if GetRemainingConstructionDuration return remaining duration time then mConstructionDuration is 100%..
(100 *  GetRemainingConstructionDuration()) / mConstructionDuration = percentage

**Reply by Vitus-TL** [reply]
Thank you, for your swift reply! Somewhere I got lost, 'cause this is basic math... Guess my head stilI spon because of trying to get grip on how this works.. I do notice that the percentage output visible on the mines has a variable lag compared to reading this value.

**Comment by AIERHANTE7619** [comment]
Hello, do you now know how to query the functions of the client? I also want to write a script, but I can't fully understand where to find the functions of the client.

**Reply by Glebsky** [reply]
Some times you can create classes from AS script, for ex. ` new (game.def("Communication.VO::dBuffVO"))` . then you debug it in introspector console. there is olso such parameters as `game`, `swmmo`, `globalFlash` etc. You can investigate already existed scripts, for futher undrestending.
